File: 238_product_of_array_except_self/productExceptSelf.go
func productExceptSelf(nums []int) []int {
	pre, pos, n := 1, 1, len(nums)
	res := make([]int, n)

	fillArray(res, 1)

	for i := range nums {
		res[i] *= pre
		pre *= nums[i]

		res[n-i-1] *= pos
		pos *= nums[n-i-1]
	}

	return res
}

func fillArray(array []int, val int) {
	for i := range len(array) {
		array[i] = val
	}
}

File: 238_product_of_array_except_self/productExceptSelf.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res;
        int pre = 1;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            res.push_back(pre);
            pre *= nums[i];
        }

        int post = 1;

        for (int i = n - 1; i >= 0; i--) {
            res[i] *= post;
            post *= nums[i];
        }

        return res;
    }
};